Add remote raiding

angelachieh opened the pull request:
This PR adds the possibility to join a raid as a remote attendee:
- New command `$remote` or `$joinremote` allows to join a raid as a remote attendee - Example: `$remote raid1`
- New command `$createremote` allows to create a raid and mark yourself (raid leader) as a remote attendee - Example: `$createremote raid1 10:00 Groudon`
- `$get` displays the total number of attendees and the total number of remote attendees (as the limits for both are different)
![image](https://user-images.githubusercontent.com/645434/95530434-c3a53400-09d5-11eb-8546-0eb4b86f00c9.png)

- New command `$mode` or `$attendance` changes the attendance mode for yourself or another user - Examples: `$mode raid1`, `$mode raid1 @user`
- New command `$invites` prints the raid's invite string, a string the leader can copy-paste in-game to easily invite all remote raiders when the raid starts  (QoL feature). This command uses nicknames instead of usernames whenever nicknames are set (users can change their nickname to match their in-game name in order to make things easier for raid leaders) - Example: `$invite raid1`
- New command `$go` or `$start` or `$lobby` notifies everyone the raid is starting (QoL feature) - Example: `$go raid1`
- `$help` output was modified because, with the new commands, the embed had more characters than Discord's limit. Now, one embed per module is built and sent.

alanschapira commented:
Just a couple of notes for myself before review

- If there are 2 modes, do we need to specify or just switch with the command `$mode raid1`

- Can anyone do `$go` or is that limited to leader?

alanschapira commented on RaidBot/BusinessLogic/Raids/RaidService.cs:
```diff
@@ -526,16 +537,98 @@ public class RaidService {
+         var raids = _raidFileService.GetRaidsFromFile().Where(a => a.Name.Equals(raidName, StringComparison.CurrentCultureIgnoreCase));
+
+         if (raids.Count() == 1) {
+            string[] validModes = { "inperson", "in-person", "remote" };
```
Since there are only 2 options, I think it could be better to not specify what you are changing to and just write `$mode raid1`

angelachieh replied:
True, it's much simpler this way! Just updated it.

alanschapira commented on RaidBot/Entities/Raid.cs:
```diff
@@ -54,13 +64,31 @@ public int UserCount
 
-            foreach (var user in Users.Skip(1)) {
-               raidWithUsers.Append($"{user.Username} {user.GetGuests} | ");
+            foreach (var user in Users) {
```
There is grouping you can do in LINQ. Not necessary, but just wanted to point that out!

angelachieh replied:
👍 Thanks!

alanschapira commented on RaidBot/Modules/RaidModule.cs:
```diff
@@ -248,8 +274,22 @@ public class RaidModule : ModuleBase {
+      [Alias("Start", "Lobby")]
+      public async Task Go([Summary("The name of the raid")] string raidName) {
+         var requesterUser = Context.User as IGuildUser;
+         var result = _raidService.MessageUsers(raidName, $"The raid is starting now. Please go to the lobby");
```
Is this something we want only the leader to be able to do?

angelachieh replied:
I thought about it, but this is simply a shortcut for `$message raidName Let's jump in`, which today anyone can do. Messaging everyone in a raid is not restricted to the leader - we can leave `$go` like this as well and see how it goes. In practice, if someone tells everyone to jump in mistakenly this will be noticed and people will probably tell the person not to do this again anyway.

angelachieh commented:
I added `$createremote` which works like `$create` but marks the raid leader as a remote attendee (note that this will only work if the server setting for auto join on raid creation is On)

I also removed the command `$create raidName` (no time, no raid boss) as nobody really uses this - also, it doesn't really make sense to create a blank raid (if it's a non-hatched raid, today we have placeholders for that, `level1`, `level5`, etc.).

alanschapira reviewed (approved):
Fantastic stuff. Thanks so much :-)
